[SE-0252][TypeChecker] Keypath Dynamic Member Lookup

xedin opened the pull request:
Implement keypath based dynamic member lookup which augments
functionality of existing @dynamicMemberLookup attribute.

Three new subscript overloads are accepted:

```swift
subscript(dynamicMember member: KeyPath<T, ...>) -> ...
subscript(dynamicmember member: WritableKeyPath<T, ...>) -> ...
subscript(dynamicMember member: ReferenceWritableKeyPath<T, ...>) -> ...
```

**Example**:

```swift
struct Point {
  let x: Int
  var y: Int
}

@dynamicMemberLookup 
struct Lens<T> {
  var obj: T

  init(_ obj: T) {
    self.obj = obj
  }

  subscript<U>(dynamicMember member: KeyPath<T, U>) -> Lens<U> {
    get { return Lens<U>(obj[keyPath: member]) }
  }

  subscript<U>(dynamicMember member: WritableKeyPath<T, U>) -> Lens<U> {
    get { return Lens<U>(obj[keyPath: member]) }
    set { obj[keyPath: member] = newValue.obj }
  }
}

var lens = Lens(Point(x: 0, y: 0))

_ = lens.x // converted into `lens[dynamicMember: KeyPath<Point, Int>`
_ = lens.y = Lens(10) // converted into `lens[dynamicMember: WritableKeyPath<Point, Int>]`
```

Resolves: rdar://problem/49069813
Resolves: rdar://problem/49533404

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[TypeChecker] Adjust dynamic member lookup to support keypaths

Allow `subscript(dynamicMember: {Writeable}KeyPath<...>)`
declarations to be found while looking for dynamic members.

xedin committed:
[TypeChecker] Add new type of overload choice to support keypath dynamic lookup

xedin committed:
[ConstraintSystem] Implement keypath dynamic member lookup

Implement keypath based dynamic member lookup which augments
functionality of existing @dynamicMemberLookup attribute.

Two new subscript overloads are accepted:

```
subscript(dynamicMember member: KeyPath<T, ...>) -> ...
subscript(dynamicmember member: WritableKeyPath<T, ...>) -> ...
```

Example:

```swift
struct Point {
  let x: Int
  var y: Int
}

@dynamicMemberLookup struct Lens<T> {
  var obj: T

  init(_ obj: T) {
    self.obj = obj
  }

  subscript<U>(dynamicMember member: KeyPath<T, U>) -> Lens<U> {
    get { return Lens<U>(obj[keyPath: member]) }
  }

  subscript<U>(dynamicMember member: WritableKeyPath<T, U>) -> Lens<U> {
    get { return Lens<U>(obj[keyPath: member]) }
    set { obj[keyPath: member] = newValue.obj }
  }
}

var lens = Lens(Point(x: 0, y: 0))

_ = lens.x // converted into `lens[dynamicMember: KeyPath<Point, Int>`
_ = lens.y = Lens(10) // converted into `lens[dynamicMember: WritableKeyPath<Point, Int>]`
```

xedin committed:
[CSApply] Implement AST transformation for keypath dynamic member lookup

xedin committed:
[CSRanking] Dynamic member look - rank keypath choice higher than string based one

Because keypath based choice provides more type information it
should be preferred to string based when both solutions are available.

xedin committed:
[TypeChecker] NFC: Add tests for keypath dynamic member lookup

xedin commented:
I think I can factor out some logic from keypath constraint so I can avoid creating implicit argument expressions, but it's vital to figuring out which overload to use `KeyPath` or `WritableKeyPath` when both are present.

xedin committed:
[ConstraintSystem] NFC: Extract check whether storage for keypath is read-only

xedin committed:
[ConstraintSystem] Add a special locator for keypath dynamic member lookup

New locator is going to store pointer to the base keypath declaration.
This information is useful to determine whether such keypath is appropriate
to access underlying properties/subscripts e.g. `WritableKeyPath` cannot be
used to access read-only storage.

xedin committed:
[CSApply] Extract keypath property resolution logic

xedin committed:
[CSSolver] Don't generate implicit argument for keypath dynamic lookup

Implicit argument expression was necessary to generate keypath
constraint which is used to validate a choice picked for the member.
But since read-only check has been factored out it's now possible
to validate choice directly in combination with new 'keypath dynamic lookup'
locator associated with member type variable which represents result
of the dynamic lookup.

xedin committed:
[ConstraintSystem] Move unviable keypath dynamic member check into `performMemberLookup`

xedin committed:
[Sema] Adjust invalid @dynamicMemberLookup to mention keypath parameter

xedin committed:
[TypeChecker] Add more tests for keypath dynamic member lookup

Test inheritance, precedence, existentials, IUO, function type
returns and more.

xedin committed:
[CSApply] Extract keypath subscript resolution logic

xedin committed:
[TypeChecker] Add subscript support keypath dynamic member lookup

tkremenek commented:
@swift-ci clean test

xedin committed:
[ConstraintSystem] Teach sanitizer about dynamic member lookup expressions

Type-check based diagnostics could introduce keypath dynamic member
expressions into AST, which have to be converted back to their
original member or subscript reference form before attempting
re-typecheck.

xedin committed:
[CSApply] Avoid reusing expressions for keypath dynamic member lookup components

In presence of type-check based diagnostics we can't mutate existing
expressions, because diagnostics could be holding references to them
directly or to expressions which point to modified ones.

xedin commented:
@swift-ci please build toolchain

xedin commented:
@swift-ci please build toolchain

xedin commented:
@swift-ci please clean test

xedin committed:
[TypeChecker] Add subscript + trailing closure tests for keypath dynamic member lookup

xedin commented:
@swift-ci please build toolchain

xedin committed:
[TypeChecker] Tighten up validation of dynamic member decls

Make sure that base decl is valid and that parameter has
a valid signature.

xedin committed:
[ConstraintSystem] Don't introduce implicit keypath choice for keypath dynamic member lookup

xedin committed:
[TypeChecker] NFC: Add more tests to keypath dynamic member lookup (tuple access + subscript/property combo)

xedin commented:
@swift-ci please build toolchain

xedin commented:
@swift-ci please build toolchain

xedin committed:
[Diagnostics] Improve mutability diagnostics related to dynamic member lookup

Since there is a way to mutate through use of writable keypath
diagnostics have to be adjusted to point to the members found
via keypath member lookup instead to using catch-all
"immutable base" diagnostic.

xedin committed:
[TypeChecker] Add `ReferenceWritableKeyPath` support to keypath dynamic member lookup

xedin committed:
[TypeChecker] Fix dynamic member declaration validation to check labels

xedin commented:
@swift-ci please build toolchain

xedin commented:
@swift-ci please build toolchain

xedin commented:
Looks like macOS toolchain job has finished but github missed notification.

xedin commented:
@swift-ci please build macOS toolchain

xedin commented:
@swift-ci please build macOS toolchain

xedin commented:
@swift-ci please build toolchain

xedin commented:
@shahmishal Looks like some of the machines are running out of disk space:

```
<unknown>:0: error: fatal error encountered during compilation; please file a bug report with your project and the crash log
16:43:18 <unknown>:0: note: IO failure on output stream: No space left on device
```

tkremenek commented:
SE-0252 has been accepted.

tkremenek commented:
@swift-ci clean test

xedin commented:
I realized that I missed one thing - allowing keypath dynamic lookup to be used inside of keypath expressions e.g. `\Lens<Point>.x` or `\Lens<Rectangle>.topLeft.x`. I'm almost done with the commit to add this support, should be ready early tomorrow.

xedin committed:
[ConstraintSystem] Allow to use keypath dynamic member lookup inside keypath expressions

xedin committed:
[ConstraintSystem] Special case keypath dynamic member lookup in disjunction filtering

Since filtering currently runs as part of the `applicable function`
constraint processing, "keypath dynamic member lookup" choice can't
be attempted in-place because that would also try to operate on that
constraint, so instead let's keep the disjunction, but disable all
unviable choices.

Resolves: rdar://problem/49533404

xedin commented:
@swift-ci please clean test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please asan test

Qbyte248 commented:
@xedin Did you consider: `let lens: Lens<Lens<Point>>`? What happens in this case? I would suspect two possibilities:
- `lens` can only access the properties of the type `Lens<Point>`
- `lens` can access both, the `Lens<Point>` and `Point` properties

xedin commented:
@Qbyte248 

> `lens` can access both, the `Lens<Point>` and Point properties

I think this is the option makes most sense, but I don't have test-case in my suite to cover that at the moment, which I'm going to add.

xedin committed:
[ConstraintSystem] Make sure that keypath dynamic member lookup could be used recursively

For example if there a structure which requires multiple implicit
dynamic member calls to get the members:

```swift
struct Point { var x, y: Int }

@dynamicMemberLookup
struct Lens<T> {
   var obj: T

   subscript<U>(dynamicMember member: KeyPath<T, U>) -> U {
     get { return obj[keyPath: member] }
   }
}

func foo(_ lens: Lens<Lens<Point>>) {
  _ = lens.x
  _ = lens.obj.x
  _ = lens.obj.obj.x
}

_ = \Lens<Lens<Point>>.x
```

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xwu commented:
> @Qbyte248
> 
> > `lens` can access both, the `Lens<Point>` and Point properties
> 
> I think this is the option makes most sense, but I don't have test-case in my suite to cover that at the moment, which I'm going to add.

This option doesn't make sense to me based on the implementation of `Lens`:

```
   subscript<U>(dynamicMember member: KeyPath<T, U>) -> U {
     get { return obj[keyPath: member] }
   }
```

The getter here does _not_ do recursive dynamic member lookup: it only uses a plain key path subscript.

Moreover, I don't see how one can even use compiler magic to synthesize this feature in the general case (i.e., for any type other than `Lens`)--how exactly are you going to recursively look up dynamic members for a type `NotALens<T, U, V>`?

xedin commented:
I think original comment referred to Lens from description which does return another Lens from dynamic lookup method.

xedin commented:
Also even if you consider subscript with returns `U` in case of `Lens<Lens<Point>>` U would be `Lens<Point>` and then `Int` so call to `lens.x` is equivalent to something like `lens[dynamicMember: \.[dynamicMember: \.x]]`

xwu commented:
> Also even if you consider subscript with returns `U` in case of `Lens<Lens<Point>>` U would be `Lens<Point>` and then `Int` so call to `lens.x` is equivalent to something like `lens[dynamicMember: \.[dynamicMember: \.x]]`

Sorry, I don’t understand. It would be like that if the getter had such an implementation (and could spell such a thing), but the getter is written in plain Swift and returns an ordinary keypath subscript that is not recursively dynamic.

xedin commented:
@xwu Let me try to clarify, dynamic subscript here has a special behavior which tries to use root type fo the KeyPath<T, U> type to lookup members so first lookup would have `Lens<Point>` as a root which would find another dynamic subscript as a member, so next lookup is going to use result of `Point` to lookup `x` on that. This whole thing results in a single keypath `KeyPath<Lens<Point>, Int>` where subscript component (for dynamic lookup) would have an argument of property component.

I have this implemented, let me try to build a toolchain so you can try this out.

xedin commented:
@swift-ci please build toolchain

xedin commented:
Source compatibility failures are related to #23140

xedin commented:
@swift-ci please ASAN test

xedin commented:
@shahmishal Looks like toolchain reason build completed but there is probably no notification about that fact to github?

xedin commented:
@swift-ci please ASAN test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
Ok, looks like all of the tests are green. I talked to @DougGregor offline and we decided to do a post-merge review of these changes, so I'm going to merge everything.

xwu commented:
> which would find another dynamic subscript as a member,

Yeah this is where I disagree—or maybe misunderstand. It should __not__ find another dynamic subscript as member.

The getter spells out the lookup as using an ordinary key path subscript, _not_ dynamic member lookup. Unless it is your intention that, on `@dynamicMemberLookup` types, use of the ordinary key path _subscript_ should also be dynamic—but that’s not a dynamic _member_ lookup but a dynamic _key path subscript_ lookup.

Put another way, when `obj` supports dynamic member lookup, `obj.member` is a dynamic lookup, but `obj[keyPath: keyPath]` shouldn’t be.

xedin commented:
I am not exactly sure why this has anything to do with `get { return obj[keyPath: member] }` call in your opinion, could you please elaborate? I am trying to point out that it is possible to construct a KeyPath argument to a single dynamic member invocation in a way that would itself have a dynamic lookup as a component to reference an underlaying property if root type of the keypath is a type which supports dynamic lookup, which is the case for `Lens<Lens<Point>>`, where root is going to be `Lens<Point>`. KeyPath expression argument is going to have a single subscript component which refers to keypath dynamic member subscript with index referring to another keypath expression with a single property component pointing to `x`.

Are you trying to say that we shouldn’t support referencing `subscript(dynamicMember:)` as a keypath component?

xwu commented:
Given `let lens:Lens<Lens<Point>>`:
* `lens.x` causes invocation of the subscript  
`lens[dynamicMember: KeyPath<Lens<Point>, Int>]` with argument `\.x`
* The subscript getter invokes the subscript  
`lens.obj[keyPath: KeyPath<Lens<Point>, Int>]`  with argument `\.x`,  
which is __not__ dynamic, and __not__ the subscript  
`lens.obj[dynamicMember: KeyPath<Point, Int>]`

In other words, I simply don't see how the getter permits `lens.x` to be equivalent to `lens[dynamicMember: \.[dynamicMember: \.x]]`; in fact, it forbids it, since the getter makes explicit that there is _no_ recursive dynamic member lookup.

xedin commented:
@xwu Maybe you should just try debug toolchain and run it with `-debug-constraints` to see how it ends up working? Because I don't know how to explain it better and I don't understand what you are trying to say. I have this example working example:

```swift
struct Point {
  let x, y: Int
}

struct Rectangle {
  var topLeft, bottomRight: Point
}

@dynamicMemberLookup 
struct Lens<T> {
  var obj: T

  init(_ obj: T) {
    self.obj = obj
  }

  subscript<U>(dynamicMember member: KeyPath<T, U>) -> U {
    get { return obj[keyPath: member] }
  }
}

func foo(_ lens: Lens<Lens<Lens<Point>>>) {
  print(lens.x)
}

var point = Point(x: 2, y: 3)
foo(Lens(Lens(Lens(point))))

var kp = \Lens<Lens<Point>>.y
var lens = Lens(Lens(point))
print(lens[keyPath: kp])
```

xwu commented:
> @xwu Maybe you should just try debug toolchain and run it with `-debug-constraints` to see how it ends up working?

I'm pretty sure I understand _how_ it's working. I don't understand how it can be regarded as consistent with the semantics of the protocol.

You are, as you say, making `lens.x` equivalent to `lens[dynamicMember: \.x]` or `lens[dynamicMember: \.[dynamicMember: \.x]]` or `lens[dynamicMember: \.[dynamicMember: \.[dynamicMember: \.x]]]`, etc., to arbitrary levels of nesting, so long as the root of the key path is itself a type that supports dynamic member lookup by key path.

But that is not what the semantics of the protocol demand. Whether `lens[dynamicMember: \.x]` is or isn't equivalent to `lens[dynamicMember: \.[dynamicMember: \.x]]` is up to the implementation of `subscript<U>(dynamicMember member: KeyPath<T, U>) -> U`.

xedin commented:
@xwu That's kind of my point, type-checker doesn't really know how `subscript<U>(dynamicMember: KeyPath<T, U>) -> U` is implemented. It only gets a declaration, opens it, and tries to combine all of the member lookups because on the rules we defined, if that leads to a solution - great! Can't really know what getter does inside from subscript use.

I'm fine saying that looking up member on `Lens<Point>` can't return in key path dynamic member lookup if current root type has itself been formed using key path dynamic lookup. I just don't understand why.
